pkg/lib/ldap/client.go

```diff
@@ -0,0 +1,106 @@
+package ldap
+
+import (
+	"errors"
+	"net/url"
+
+	"github.com/go-ldap/ldap/v3"
+
+	"github.com/authgear/authgear-server/pkg/api"
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/authgear/authgear-server/pkg/util/ldaputil"
+)
+
+type Client struct {
+	Config config.LDAPServerConfig
+	conn   *ldap.Conn
```

Comment:
First of all, a Client should be stateless. It does not contain a conn.

A common pattern in programming with resources is to return a handle. https://en.wikipedia.org/wiki/Handle_(computing) In this case, you return an interface to represent the handle (that is, the connection)

In this case, we expect this client to do one simple thing, which is to authenticate an end-user, so it should expose only one method.

```golang
type Client struct {
  Config config.LDAPServerConfig
}

func (c *Client) AuthenticateEndUser(username string, password string) (*ldap.Entry, error) {
  // Call the private method to make a new connection.
  conn, err := c.makeConnection()
  if err != nil {
    return nil, err
  }
  // Use defer statement to close the connection.
  defer conn.Close()

  // Call the private method to bind the search user.
  err := c.bind()
  if err != nil {
    return nil, err
  }
  
  entry, err := c.authenticateEndUser(username, password)
  if err != nil {
    return nil, err
  }
  
  return entry, nil
}
```

Reply:
Check if unique attribute exist in entry